### PR TITLE
DADS: Unsupervised Reinforcement Learning

### PR DESCRIPTION
**Pull request by AntoineTheb**
~Not ready for review yet~
Donesies

### CODE REVIEW

File: article/_posts/2020-07-30-DADS.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+
+# Highlights
+
+- Unsupervised "skill" discovery without extrinsic rewards
+- Planning in "skill" space allows to generalize to unseen tasks
+
+# Introduction
+
+Model-based reinforcement learning (MBRL) tries to learn the dynamics of the environment as well as a policy that plans the optimal course of action using the learned dynamics. This allows the polci to be learned in more sample-efficient way as well as making able to generalize to new tasks.
```
Comment (jodoin) [inline article/_posts/2020-07-30-DADS.md:21]: polci -> policy

File: article/_posts/2020-07-30-DADS.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+However, generalizing to new state distributions remains an open problem, as exploring the full state-space and learning the full dynamics for high-dimensional systems is generally infeasible. The authors want to retain the flexibility of MBRL by using the simplicity of model-free RL to learn low-level behaviors than can be leveraged at test time.  
+
+# Methods
+
+The authors want to learn the dynamics for a large and diverse set of skills, which allows them to to use model-based planning in the _skill space_ for solving new tasks.
```
Comment (jodoin) [inline article/_posts/2020-07-30-DADS.md:27]: to to

File: article/_posts/2020-07-30-DADS.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Unsupervised skill discovery
+
+Using the common MDP $$M = (S, A, p, r)$$, they want to learn a skill-conditioned policy $$\pi(a \vert s,z)$$ with $$z \in \mathbb(Z)$$ than can be sampled from a prior $$p(z)$$. To plan in the _skill-space_, the authors also propose to learn the dynamics for skills $$q_{\theta}(s'{\vert}s,z)$$, which will be used to plan at test time.
```
Comment (jodoin) [inline article/_posts/2020-07-30-DADS.md:31]: that can

File: article/_posts/2020-07-30-DADS.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+with $$L$$ the number of samples drawn from $$p(z)$$. Again, the reward function encourages the policy to produce transitions that are predictable under $$q_\theta$$ as well as diversified.
+
+In implementation, the authors use the Soft Actor-Critic algorithm to train their policy, and their prior $$p(z) = U(-1, 1)^D$$ where $$D$$ is the dimentionality of the input space.
+
+## Planning in action space
+
+## Data
+
+
+# Results
```
Comment (jodoin) [inline article/_posts/2020-07-30-DADS.md:64]: Results?

File: article/_posts/2020-07-30-DADS.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+However, generalizing to new state distributions remains an open problem, as exploring the full state-space and learning the full dynamics for high-dimensional systems is generally infeasible. The authors want to retain the flexibility of MBRL by using the simplicity of model-free RL to learn low-level behaviors that can be leveraged at test time.  
+
+# Methods
+
+The authors want to learn the dynamics for a large and diverse set of skills, which allows them to use model-based planning in the _skill space_ for solving new tasks.
+
+## Unsupervised skill discovery
+
+Using the common MDP $$M = (S, A, p, r)$$, they want to learn a skill-conditioned policy $$\pi(a \vert s,z)$$ with $$z \in \mathbb(Z)$$ that can be sampled from a prior $$p(z)$$. To plan in the _skill-space_, the authors also propose to learn the dynamics for skills $$q_{\theta}(s'{\vert}s,z)$$, which will be used to plan at test time.
```
Comment (jhlegarreta) [inline article/_posts/2020-07-30-DADS.md:31]: Although not very important, I'd stick to the same convention when mentioning the _skill space_: i.e. with or without a dash, but always with the same form.

File: article/_posts/2020-07-30-DADS.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+![](/article/images/DADS/fig2.jpeg)
+
+To discover skills, the authors propose to maximise the mutual information (MI) between the next state $$s'$$ and the skill $$z$$ conditioned on the current state $$s$$ s.t.
+
+![](/article/images/DADS/eq1-2.jpeg)
+
+Maximizing the MI corresponds to maximizing the diversity of skills in the latent space $$Z$$ (therefore maximizing the relevancy of skills) while making the transitions for a given $$z$$ predictable (thus making the learning process stable).
+
+Because the actual dynamics are unknown but can be sampled, maximizing (2) is intractable, but the objective can atleast be lower-bounded by
```
Comment (jhlegarreta) [inline article/_posts/2020-07-30-DADS.md:41]: _atleast_ -> _at least_

**Comment by AntoineTheb** [comment]
@jhlegarreta Good point, I included a mention on the usage of behavior/skill space. Also, I initially did not include the part on hierarchical RL because I'm not too knowlegable on HRL but you're right I should have mentionned it. I also realized that the figure also shows performance against model-free RL so that's a plus.

File: article/_posts/2020-07-30-DADS.md (hunk 5 above)
Comment (jhlegarreta) [reply article/_posts/2020-07-30-DADS.md]: Addressed in 4fc7473.

File: article/_posts/2020-07-30-DADS.md (hunk 6 above)
Comment (jhlegarreta) [reply article/_posts/2020-07-30-DADS.md]: Addressed in 4fc7473.

**Comment by lemairecarl** [comment]
Two questions:

1. I don't understand how maximizing mutual information between s and s' is maximizing the diversity of the behaviors. Can you explain, please? Or point something to read.

2. I understand that the method plans by constructing a sequence of states using q(s' | s, z), and that the reward can be computed from s. Is s a latent vector describing the state? If so, I'm not sure how you can get the reward from that. With the value function?

**Comment by AntoineTheb** [comment]
> I don't understand how maximizing mutual information between s and s' is maximizing the diversity of the behaviors. Can you explain, please? Or point something to read.
MI is actually maximized between s' and z, conditionned on s, so that z explains s' given s. I can point you to line 35 of my review 😬 As to why it's promoting the diversity of behaviors, again I can point you to equation (1) and (2) of the review, where the relation between the MI and the entropy H(s'|s) is shown. Maximizing H(s'|s) means promoting diverse transitions, but equation 2 constrains the diverse transitions to be explainable by z, otherwize H(s'|s,z) would also be high and thus the subtraction would give a low score. I agree this can be criptic from just the equations, I'll clarify

> I understand that the method plans by constructing a sequence of states using q(s' | s, z), and that the reward can be computed from s. Is s a latent vector describing the state? If so, I'm not sure how you can get the reward from that. With the value function?

No, s follows standard RL notation and thus represents the state of the environment. $$q(s'|s,z)$$ are the learned dynamics of the environment, conditioned on the latent variable. So the model does not actually learn the whole environment dynamics but only the subset that pertains to the skills learned. I'll clarify. Because s does represent the state, you can feed it to reward function as well as a candidate action coming from your behaviors to get the actual environment reward. I'll clarify that also.

**Comment by lemairecarl** [comment]
@AntoineTheb Thanks! I understand now.

**Pull request merged**